docs/rfcs/2024-04-01-super-peers-2.md

```diff
@@ -0,0 +1,262 @@
+-- by default all super peers will deliver the events in that scope.
+data GroupRedundancy = GroupRedundancy
+  { messages :: Maybe Double, -- messages and message changes, including reactions and comments
+    members :: Maybe Double, -- member additions and permission changes
+    group :: Maybe Double -- group profile and other changes
+  }
+```
+
+## Decisions about message or another event delivery
+
+In groups with rank 0 (current groups) all members aim to establish connections with all other members, making it hard to scale. Admins who added the members play a temporary role of forwarding messages, but only until members are connected.
+
+In groups with rank 1 super-peers connect to all members, but each message or event may be delivered by some rather than by all super-peers, to avoid substantial traffic increase. E.g., in groups with 2 super-peers it would be desirable to deliver each message 1.33 (4/3) times on average, while for groups with 3 super peers it can be 2 or 1.667 (5/3) times.
+
+The decision whether a given super peer should deliver the message would depend on these factors:
+- deterministic (see below) message hash, `h`.
+- receiving member ID, `r`.
+- super-peer member ID, `s`.
+- target message redundancy `d` (the desired number of super-peers to deliver the message).
+- number of super-peers connected to member `r` (and known as connected to super-peer making the decision), `n`.
+- 0-based index of the current super-peer in the sorted array of known super peer IDs, `i`. It does not require sorting all peer IDs, it's enough to count how many peers have smaller or larger IDs.
+
+The assumption here is that member ID is unique within the group, and that message hashes are also unique. Also, message hashes rather than sending member IDs are used to ensure that the decision is made differently for the same sender/recipient pairs, allowing recipients to identify integrity violations (e.g., if some of the super-peers decides to change the messages or fail to deliver it).
+
+For simplicity, all parameters are normalized to 0..1 range from their respective binary ranges.
+
+The algorithm to decide whether the message should be delivered by a given super-peer:
+
+```
+if (d >= n || n == 1) deliver;
+else
+  prob = d/n ; delivery probability based on target delivery redundancy
+  point = (h + r)/2 ; as member ID and message hash are uniformly random in 0..1 range, `point` will also be uniformly random in 0..1 range
+  start = (1 - prob) * i / (n - 1) ; `start` for the peer with `i == 0` will be `0` and with `i == n - 1` will be `1 - prob`
+  end = start + prob ; `end` for peer 0 will be `prob` and for peer `n - 1` will be 1
+  ; for all peers the range `start..end` will have width `prob`
+  if (point >= start && point < end) deliver
+  else skip;
+```
+
+This algorithm can be proven to result in target average delivery redundancy.
```

Comment:
> E.g., in groups with 2 super-peers it would be desirable to deliver each message 1.33 (4/3) times on average, while for groups with 3 super peers it can be 2 or 1.667 (5/3) times.

so in practice the messages will have different redundancy? Why not set redundancy to integers, and require super peers to aim for it for each message, seems like it would make more sense than arbitrarily passing some messages without redundancy. Even in same scope of messages some messages can be more important than others, it doesn't seem correct to leave these decisions to chance / rely on requests for re-delivery (?). Even if they're implemented, errors (failed deliveries) are likely in network and it might incur comparable traffic to simply having redundancy set at least to 2 and higher complexity of protocol.

Super-peers could calculate the "index" of a given message, with peers to deliver it determined based on some rule, for example "index" peer and next in circle.